### PR TITLE
Add in-place Stats in `cg`

### PR DESCRIPTION
**Pull request by tmigot**
@amontoison toward 0-allocation cg :).
Let me know if you have any comments and I will keep going with the others, #361 .

### CODE REVIEW

File: src/cg.jl [hunk 1]
```diff
@@ -64,23 +64,28 @@ function cg!(solver :: CgSolver{T,S}, A, b :: AbstractVector{T};
...
   x .= zero(T)
   r .= b
   MisI || mul!(z, M, r)
   p .= z
   γ = @kdot(n, r, z)
-  γ == 0 && return x, SimpleStats(true, false, [zero(T)], T[], "x = 0 is a zero-residual solution")
+  if γ == 0 
+    stats.solved, stats.inconsistent = true, false
+    stats.status = "x = 0 is a zero-residual solution"
+    return x, stats
```
Comment (dpo) [inline src/cg.jl:79]: It's enough to return the solver in all cases, isn't it?

File: src/krylov_solvers.jl [hunk 2]
```diff
@@ -76,7 +77,8 @@ mutable struct CgSolver{T,S} <: KrylovSolver{T,S}
     p  = S(undef, n)
     Ap = S(undef, n)
     z  = S(undef, 0)
-    solver = new{T,S}(x, r, p, Ap, z)
+    stats = SimpleStats(false, true, T[], T[], "unknown")
+    solver = new{T,S}(x, r, p, Ap, z, stats)
```
Comment (dpo) [inline src/krylov_solvers.jl:81]: Maybe solvers could use a few getters: `solution(solver) = solver.x`, `stats(solver) = solver.stats`, ...

File: src/cg.jl (hunk 1 above)
Comment (tmigot) [reply src/cg.jl:79]: Sounds like a good idea. Maybe `cg!` returns the solver, but `cg` returns `x, stats` as it is now?

Comment (dpo) [reply src/cg.jl:79]: They should be consistent and always return the solver in my opinion.

Comment (amontoison) [reply src/cg.jl:79]: We can only return the solver after #361 but we need to modify all solvers in the same time.

File: src/krylov_solvers.jl (hunk 2 above)
Comment (amontoison) [reply src/krylov_solvers.jl:81]: Good idea ! I propose to do it after #361 :smiley:

File: src/cg.jl [hunk 3]
```diff
@@ -64,23 +64,28 @@ function cg!(solver :: CgSolver{T,S}, A, b :: AbstractVector{T};
 
   # Set up workspace.
   allocate_if(!MisI, solver, :z, S, n)
-  x, r, p, Ap = solver.x, solver.r, solver.p, solver.Ap
+  x, r, p, Ap, stats = solver.x, solver.r, solver.p, solver.Ap, solver.stats
+  rNorms = solver.stats.residuals
```
Comment (amontoison) [inline src/cg.jl:68]:
```suggestion
  rNorms = solver.stats.residuals
  !history && !isempty(rNorms) && (rNorms = T[])
```

File: src/cg.jl [hunk 4]
```diff
@@ -64,23 +64,29 @@ function cg!(solver :: CgSolver{T,S}, A, b :: AbstractVector{T};
...
+  !history && !isempty(rNorms) && (rNorms .= T[])
```
Comment (amontoison) [inline src/cg.jl:69]:
```suggestion
  !history && !isempty(rNorms) && (rNorms = T[])
```

Comment (tmigot) [reply src/cg.jl]: Thanks

**Comment by amontoison** [comment]
Thanks Tangi !

**Pull request merged**